src/ring_theory/unique_factorization_domain.lean

```diff
@@ -455,3 +455,103 @@ def unique_factorization_domain.to_gcd_domain
   .. ‹normalization_domain α› }
 
 end
+
+namespace unique_factorization_domain
+
+variables {R : Type*} [integral_domain R] [unique_factorization_domain R]
+
+lemma no_factors_of_no_prime_factors {a b : R} (ha : a ≠ 0)
+  (h : (∀ {p}, prime p → p ∣ a → p ∣ b → is_unit p)) : ∀ {p}, p ∣ a → p ∣ b → is_unit p :=
+λ p, induction_on_prime p
+  (by { simp only [zero_dvd_iff], intros, contradiction })
+  (λ x hx _ _, hx)
+  (λ p q hp hq ih dvd_a dvd_b, is_unit.mul
+    (h hq (dvd_of_mul_right_dvd dvd_a) (dvd_of_mul_right_dvd dvd_b))
+    (ih (dvd_of_mul_left_dvd dvd_a) (dvd_of_mul_left_dvd dvd_b)))
+
+lemma left_dvd_or_dvd_right_of_dvd_prime_mul {a : R} :
+  ∀ {b p : R}, prime p → a ∣ p * b → p ∣ a ∨ a ∣ b :=
+begin
+  refine induction_on_prime a _ _ _,
+  { intros b p _ ha,
+    refine (eq_zero_or_eq_zero_of_mul_eq_zero (zero_dvd_iff.mp ha)).imp _ _;
+      rintro ⟨rfl⟩; refl },
+  { intros x x_is_unit b _ _ _,
+    exact or.inr (is_unit_iff_forall_dvd.mp x_is_unit b) },
+  { intros a q a_ne_zero hq ih b p hp qa_dvd,
+    cases (hq.div_or_div (dvd_of_mul_right_dvd qa_dvd)) with q_dvd_p q_dvd_b,
+    { left,
+      apply dvd_mul_of_dvd_left,
+      refine dvd_symm_of_irreducible (irreducible_of_prime _) (irreducible_of_prime _) _;
+        assumption },
+    { rcases q_dvd_b with ⟨b', rfl⟩,
+      rw mul_left_comm at qa_dvd,
+      refine (ih hp ((mul_dvd_mul_iff_left hq.ne_zero).mp qa_dvd)).imp _ _,
+      { exact λ h, dvd_mul_of_dvd_right h _ },
+      { exact mul_dvd_mul_left q } } }
+end
```

Comment:
The UFD assumption here is unnecessary. Here's a proof for an integral domain.

```suggestion
begin
  rintros b p hp ⟨c, hc⟩,
  rcases hp.2.2 a c (hc ▸ dvd_mul_right _ _) with h | ⟨x, rfl⟩,
  { exact or.inl h },
  { rw [mul_left_comm, domain.mul_right_inj hp.ne_zero] at hc,
    exact or.inr (hc.symm ▸ dvd_mul_right _ _) }
end
```

Reply:
I have added your suggestion to `ring_theory.integral_domain`. Thanks!